### PR TITLE
MAINT: randomly sample chunks for partial_fit calls

### LINKED ISSUE #274: Incremental.partial_fit does not randomly shuffle blocks on repeated calls
Or blocks are always fed into `estimator.partial_fit` in the same order. Here's an example of repeated `partial_fit` calls (interspersed by `score` calls):

screen shot 2018-07-03 at 11 29 11 am

I am sure this does not matter for the test data we create, but it might matter for other data (e.g., time series).

### PR DESCRIPTION
**Pull request by stsievert**
Closes #274.

A screenshot of different `partial_fit` calls:

| Before (almost) | After |
|:--:|:--:|
|screen shot 2018-07-03 at 11 06 04 am | screen shot 2018-07-03 at 1 07 36 pm |

I only show the before for some comparison, though it's not exact (and was in the debug process). It is "almost" because it uses a fixed seed every time, not a random seed. It does not run through the blocks in sequential order as mentioned in the previous docs.

### CODE REVIEW

**Comment by stsievert** [comment]
I've had to reformat `test_incremental_basic` to train longer and assert that the results are close enough, not exactly the same.

File: dask_ml/_partial.py [hunk 1]
```diff
@@ -171,12 +172,15 @@ def fit(model, x, y, compute=True, **kwargs):
         x = x.rechunk(chunks=(x.chunks[0], sum(x.chunks[1])))
 
     nblocks = len(x.chunks[0])
+    order = list(range(nblocks))
+    random.shuffle(order)
```
Comment (TomAugspurger) [inline dask_ml/_partial.py]: For reproducibility, Incremental should take a `random_state` and then use that for shuffling the data.

I also think the shuffling should be optionally enabled / disabled (not sure what the default should be) by a hyper parameter.

Finally, I'm not sure what the parameter name should be. Typically scikit-learn uses `shuffle=True/False`. We can either re-use that, or call it `shuffle_blocks`, and reserve `shuffle` for when we actually shuffle data between blocks.

Comment (stsievert) [reply dask_ml/_partial.py]: I think shuffling should be enabled by default. That's what sklearn default too, and it's what SGD theory points too.

I've used `shuffle_blocks`; that describes exactly what it does. We can use `shuffle` for full shuffling, and `shuffle_each_block` to shuffle in each block.

**Comment by stsievert** [comment]
I have actually tested this now, and believe it's ready for merge.

**Comment by mrocklin** [comment]
This all seems fine to me.  The sklearn-dev failures are interesting.  Do we know if it is due to this PR or something upstream?

**Comment by stsievert** [comment]
I don't see how it could be this PR. All the failures are with `RobustScaler`, which doesn't call `fit` or `Incremental`.

**Comment by mrocklin** [comment]
OK.  Fixing it separately here: https://github.com/dask/dask-ml/pull/283

**Comment by stsievert** [comment]
There was a PR on `RobustScalar` that was merged earlier this morning: https://github.com/scikit-learn/scikit-learn/pull/11308.

**Comment by mrocklin** [comment]
That would help explain things :)

File: tests/test_incremental.py [hunk 2]
```diff
@@ -6,9 +6,9 @@
 from dask.array.utils import assert_eq
 from sklearn.base import clone
 from sklearn.linear_model import SGDClassifier
+import numpy.linalg as LA
```
Comment (mrocklin) [inline tests/test_incremental.py]: The `LA` pronoun isn't very common.  I'd prefer the fully qualified name if possible.

```python
import numpy as np


np.linalg.svd(...)
```

File: dask_ml/wrappers.py [hunk 3]
```diff
@@ -323,6 +323,12 @@ class Incremental(ParallelPostFit):
            a single NumPy array, which may exhaust the memory of your worker.
            You probably want to always specify `scoring`.
 
+    random_state : int or numpy.random.RandomState
```
Comment (TomAugspurger) [inline dask_ml/wrappers.py]: Append `, optional`.

File: dask_ml/wrappers.py [hunk 4]
```diff
@@ -323,6 +323,12 @@ class Incremental(ParallelPostFit):
...
+        Random object that determines how to shuffle blocks.
+
+    shuffle_blocks : bool
```
Comment (TomAugspurger) [inline dask_ml/wrappers.py]: append `, default True`.

File: dask_ml/wrappers.py [hunk 5]
```diff
@@ -323,6 +323,12 @@ class Incremental(ParallelPostFit):
...
+        Whether to randomly shuffle the blocks or now
```
Comment (TomAugspurger) [inline dask_ml/wrappers.py]: typo: now -> not.

Could you elaborate on what this means? Specifically, clarify shuffling the block order from shuffling within blocks (from shuffling between blocks).

**Comment by stsievert** [comment]
Thanks for the review @mrocklin and @TomAugspurger. I've addressed those comments.

**Comment by TomAugspurger** [comment]
Thanks!
